### DISCUSSION TITLE
Setting up React Devtools (standalone) with NextJS App

### DISCUSSION THREAD

**Discussion by shufflerAbhi** [vercel/next.js#50665] [Help] [answered]
### Summary

I'm quite new to NextJS. I wish to use React Devtools for debugging my app. My organization does not allow browser extensions to be installed. I was wondering if I'd be able to use the standalone version of react-devtools (https://www.npmjs.com/package/react-devtools) to work with my Next JS App.

I created the app as a TypeScript project using the create-next-app. For simplicity, I have a very simple structure. I have only one page (index.tsx) which renders my root component 'AppRoot'. I want this component to be rendered client side, so I'm using dynamic import. It renders one child component AppContent.

When I run the standalone react-devtools using `npx react-devtools`, it wasn't connecting to the app. The readme for react-devtools mentions a couple of options to debug apps that use React DOM. Either add script tag (`<script src="http://localhost:8097"></script>`) or to install react-devtools as a dependency and import it in the app. I wasn't sure if this applied to NextJS apps and I wasn't sure where I should add the script or import the react-devtools package. I tried putting the import or the script in both index.tsx and in AppRoot.tsx individually. Devtools was able to connect to the app and show the component tree, but it was showing a timeout when I inspect the components.

![image](https://github.com/vercel/next.js/assets/19413987/8c0dac5c-0861-4e84-bdb3-b290feb22635)

My questions are

1. Can I use the standalone version of React-devtools to debug the **client side rendered** code in my NextJS App?
2. If so, how do I set this up?

pages/index.tsx
```import dynamic from "next/dynamic"
import Head from "next/head"

const AppRoot = dynamic(() => import("@/components/AppRoot"), {
  ssr: false,
})

export default function Main() {
  return (
    <>
      <Head>
        <title>Main</title>
        <meta name="description" content="Main" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <main>
        <AppRoot />
      </main>
    </>
  )
}
```

pages/_app.tsx
```
import "@/styles/globals.css"
import type { AppProps } from "next/app"

export default function App({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />
}
```

pages/_document.tsx

```
import { Html, Head, Main, NextScript } from "next/document"

export default function Document() {
  return (
    <Html lang="en">
      <Head />
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  )
}
```

components/AppRoot.tsx

```
const AppRoot = () => {
  return <AppContent />
}

export default AppRoot


const AppContent = () => {
  return <div> Hello World!</div>
```

### Additional information

```js
"next": "13.4.3"
"react": "18.2.0"
"react-devtools": "^4.27.8"
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I think you just need to place it in the `_document` file, like so:

```jsx
import { Html, Head, Main, NextScript } from "next/document";

export default function Document() {
  return (
    <Html lang="en">
      <Head>
        <script src="http://localhost:8097"></script>
      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  );
}
```

You'll see a warning/error: `Synchronous scripts should not be used. See: https://nextjs.org/docs/messages/no-sync-scriptseslint[@next/next/no-sync-scripts](https://nextjs.org/docs/messages/no-sync-scripts)`, but you can ignore that since this is for dev mode only, you should not include this in production!

I used this Home page component:

```jsx
import { useState, useDebugValue } from "react";

const useValue = () => {
  const [value, setValue] = useState(0);

  useDebugValue(value, (v) => `Hook value: ${v}`);

  return [value, setValue];
};

export default function Home() {
  const state = useValue();
  return (
    <section>
      <p>123</p>
    </section>
  );
}
```

And got this:

![Screenshot 2023-06-02 at 09 24 16](https://github.com/vercel/next.js/assets/21013447/30bd3485-852d-4eb1-b2db-eb9bcb3b4f29)

**Reply by shufflerAbhi** [reply]
Awesome! 

Thank you for the reply. I didn't think of this because I was under the impression that _dcoument is rendered on the server and the <script> might not reach the browser.

**Reply by vegerot** [reply]
When I do this I get HSTS and CORS errors.  I was wondering if it's possible to import this script from the same origin.  In other words, for somewhere in my app to say

```js
import "react-devtools";
```

It's tricky because this couldn't run on the server, but would have to be the first thing to run on the client.  idk if that's possible.  Maybe some webpack hackery?
